### ISSUE TITLE
Strange Compatibility Issue with Create and JEI

### ISSUE THREAD

**Issue by TheJacksonOfGreen** [mortuusars/Exposure#14]
When installed alongside Create and JEI in 1.20.1, the game crashed when I attempted to view the crafting recipe for developing Color Film using Create's Mechanical Crafters. Nothing game-breaking by any means, and I know you can't expect every mod to be perfectly compatible with every single other one, but I just thought I'd let you know.

**Comment by mortuusars** [comment]
Please provide the crash log.

**Comment by TheJacksonOfGreen** [comment]
oh, yes, of course. Sorry about that.

[It's right here.](https://pastebin.com/3pjsxy6k)

Very likely might be a Create or JEI bug, I don't know if there are any recipes quite like how film development works.

**Comment by krokots** [comment]
Same bug, using JEI to view automated crafting for developed films & photographs crashes Minecraft.
https://pastebin.com/3rRtz3n7

**Comment by dnk1234567** [comment]
I think this is a bug(or maybe a design not good enough?).
The film developing recipe and photograph cloning recipe implement the `IShapedRecipe<CraftingContainer>` interface and will be register to Create's auto crafter JEI page. And they provide ingredients as shapeless recipe. 
But create JEI page parses them as shaped recipes. 
```     
	private static int getWidth(CraftingRecipe recipe) {
		return recipe instanceof IShapedRecipe<?> ? ((IShapedRecipe<?>) recipe).getRecipeWidth() : 1;
	}

	private static int getHeight(CraftingRecipe recipe) {
		return recipe instanceof IShapedRecipe<?> ? ((IShapedRecipe<?>) recipe).getRecipeHeight() : 1;
	}
```
```
public void draw(CraftingRecipe recipe, IRecipeSlotsView iRecipeSlotsView, GuiGraphics graphics, double mouseX,
		double mouseY) {
		PoseStack matrixStack = graphics.pose();
		matrixStack.pushPose();
		float scale = getScale(recipe);
		matrixStack.translate(getXPadding(recipe), getYPadding(recipe), 0);

		for (int row = 0; row < getHeight(recipe); row++)
			for (int col = 0; col < getWidth(recipe); col++)
				if (!recipe.getIngredients()
					.get(row * getWidth(recipe) + col)
					.isEmpty()) {
					matrixStack.pushPose();
					matrixStack.translate(col * 19 * scale, row * 19 * scale, 0);
					matrixStack.scale(scale, scale, scale);
					AllGuiTextures.JEI_SLOT.render(graphics, 0, 0);
					matrixStack.popPose();
				}
```
But ingredents are provided as shapeless recipe.
And the recipes' width and height are set to 3.
```
    @Override
    public int getRecipeWidth() {
        return 3;
    }

    @Override
    public int getRecipeHeight() {
        return 3;
    }
```
So It will cause a `IndexOutOfBoundsException`.
I fixed them by changing the constants of width and height.(e.g. 1 * 2, 2 * 2, 1 * 3, 2 * 3) 
But they are shapeless recipes, so if you change the json the constants need to be recalculated and some conditions even not have a solution.(e.g. 7 ingredients)

![5](https://github.com/mortuusars/Exposure/assets/86105014/3dffa431-a414-4d5a-bf97-605dc494f595)

In fact, the create's auto crafter will not give you the remains, so you will lose 3 glass bottles.
Even worse, the photograph cloning recipe in auto crafter is completely useless......

**Comment by mortuusars** [comment]
It probably broke when I changed recipe width/height from 0 to 3. (because it printed error messages in logs when was set to 0. Maybe it should have been 1 from the start, idk).
I should have fixed this right away, but postponed it due to rewriting the mod to support both forge and fabric.

This whole system is a mess. IShapedRecipe is a hack to make it work with Mechanical Crafters but not with Shapeless Mixing. And the recipe using nbt ingredients is an issue in of itself.

**Comment by mortuusars** [comment]
Fixed in 1.2.0
